### PR TITLE
Layerscape LS1028ARDB/LS1046AFRWY/LX2160ARDB support

### PR DESCRIPTION
**Pull request by yangbolu1991**
The pull request patches are from NXP Layerscape OpenWrt 20.09 release. (Reworked and cleaned up)
link: https://source.codeaurora.org/external/qoriq/qoriq-components/openwrt/
tag: layerscape-20.09

The pull request cherry-picked only BSP level changes of layerscape-20.09.
1. For layerscape user space packages patches, will upstream later to package.git.
2. Some sysupgrade fix-up not cherry-picked, since below pull request also included the fix-up.
But either of the two pull request has to rebase once one is merge.
https://github.com/openwrt/openwrt/pull/3345

This pull request is mainly to update layerscape components to latest LSDK release (LSDK-20.04-update-290520),
and to add three new boards LS1028ARDB/LS1046AFRWY/LX2160ARDB.

Please help to review and merge.
Thanks a lot.

### CODE REVIEW

File: package/firmware/layerscape/ls-dpl/Makefile
Comment (adschm) [inline package/firmware/layerscape/ls-dpl/Makefile:46]: Please bump PKG_RELEASE in cases where you add stuff without changing upstream version.

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (adschm) [inline target/linux/layerscape/base-files/etc/board.d/02_network:17]: Please don't use wildcards for board names in base-files. This makes it impossible to grep for a specific device string.

File: package/firmware/layerscape/ls-rcw/Makefile
Comment (adschm) [inline package/firmware/layerscape/ls-rcw/Makefile:12]: PKG_RELEASE is the OpenWrt-internal version.
That means if you update PKG_VERSION, PKG_RELEASE goes back to "1" and is then incremented for "local" changes, until upstream is updated again ...

**Comment by adschm** [comment]
@yangbolu1991 

Can I pick only these four patches for now, or is there a dependency I'm missing:

```
layerscape: update kernel to LSDK-20.04-V5.4-update-290520
layerscape: update ls-rcw to LSDK-20.04-update-290520
layerscape: update u-boot to LSDK-20.04-update-290520
layerscape: update tfa to LSDK-20.04-update-290520
```

**Comment by adschm** [comment]
@CHKDSK88 FYI.

**Comment by adschm** [comment]
Consider submitting "perf: add libcap dependency" separately, either in a separate PR or via the mailing list. I fear it otherwise won't reach the correct people to review it.

**Comment by adschm** [comment]
I've merged the sysupgrade PR and took the four update-290520 patches plus the NXP linux factory changes from this PR.

Note that I removed the lx2160a changes from ls-rcw patch, as that device is not merged yet. Take care to include the updated patchs in the lx2160a support patch when rebasing.

File: package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-sdboot-uEnv.txt
Comment (CHKDSK88) [inline package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-sdboot-uEnv.txt:3]: Please fix ```fdt_high``` value for proper work fit image.

File: package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-sdboot-uEnv.txt
Comment (CHKDSK88) [inline package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-sdboot-uEnv.txt:3]: Please fix fdt_high value for proper work fit image.

**Review by CHKDSK88 — COMMENTED** [review]
I will look again into this PR after rebase to master.

File: package/firmware/layerscape/ls-dpl/Makefile
Comment (yangbolu1991) [reply package/firmware/layerscape/ls-dpl/Makefile:46]: Ok.

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (yangbolu1991) [reply target/linux/layerscape/base-files/etc/board.d/02_network:17]: Ok.

File: package/firmware/layerscape/ls-rcw/Makefile
Comment (yangbolu1991) [reply package/firmware/layerscape/ls-rcw/Makefile:12]: Thanks. Understand now.

**Comment by yangbolu1991** [comment]
> 
> 
> Consider submitting "perf: add libcap dependency" separately, either in a separate PR or via the mailing list. I fear it otherwise won't reach the correct people to review it.

Thanks. Will send it separately.

**Comment by yangbolu1991** [comment]
> 
> 
> I've merged the sysupgrade PR and took the four update-290520 patches plus the NXP linux factory changes from this PR.
> 
> Note that I removed the lx2160a changes from ls-rcw patch, as that device is not merged yet. Take care to include the updated patchs in the lx2160a support patch when rebasing.

Thank you for cherry-picking. I will rebase the rest soon.

File: package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-sdboot-uEnv.txt
Comment (yangbolu1991) [reply package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-sdboot-uEnv.txt:3]: Ok. Thanks.

File: package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-sdboot-uEnv.txt
Comment (yangbolu1991) [reply package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-sdboot-uEnv.txt:3]: Ok, thanks.

**Comment by yangbolu1991** [comment]
Pushed v2 patch-set.
Changes for v2:
- Rebased on latest master branch.
- Updated PKG_RELEASE for package internal changes.
- Dropped wildcards usage for board name in script.
- Dropped perf package patch and sent it separately.

Thanks!

File: package/firmware/layerscape/ls-ddr-phy/Makefile
Comment (CHKDSK88) [inline package/firmware/layerscape/ls-ddr-phy/Makefile:1]: SDPX Header?

File: package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-sdboot-uEnv.txt
Comment (CHKDSK88) [reply package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-sdboot-uEnv.txt:3]: My two boards (LS1012A-FRDM and LS1046A-RDB) won't start kernel from fit image with `fdt_high=0xffffffffffffffff`. Could You confirm, that it works for this boards?

File: package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-sdboot-uEnv.txt
Comment (CHKDSK88) [reply package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-sdboot-uEnv.txt:3]: Are You sure, that this board boot with `fdt_high=0xffffffffffffffff`?

File: package/firmware/layerscape/ls-ddr-phy/Makefile
Comment (yangbolu1991) [reply package/firmware/layerscape/ls-ddr-phy/Makefile:1]: Let me add it.

**Comment by yangbolu1991** [comment]
@CHKDSK88 
For the fdt_high setting, during my previous tests, booting was fine.
But I just double checked, I indeed got booting issue. This happened when the dtb was loaded at address which was not 8-byte boundary.

So thanks for your reminding. We may have several ways to resolve, like setting proper fdt_high, or using bootm_size which may be better. And this may be also resolved by making kernel its specifying loadaddr for fdt (It seems current openwrt script does not support). Anyway, I will add the fix in new version patches.
Thanks.

File: package/firmware/layerscape/ls-ddr-phy/Makefile
Comment (dangowrt) [inline package/firmware/layerscape/ls-ddr-phy/Makefile:18]: PKG_LICENSE?

**Review by dangowrt — CHANGES_REQUESTED** [review]
`PKG_LICENSE` is missing for all packages added. Please add that as well as `PKG_LICENSE_FILES` (if any).

**Comment by CHKDSK88** [comment]
@yangbolu1991 

> Anyway, I will add the fix in new version patches.

Thank You. I will wait for new version of PR.

**Comment by yangbolu1991** [comment]
Pushed v3 patch-set.
Changes for v3:
- Added SPDX header and PKG_LICENSE/PKG_LICENSE_FILES for new ls-ddr-phy package.
- Fixed SD booting issue by using bootm_size for new boards ls1046afrwy/ls1028ardb/lx2160ardb.

Thanks.

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (CHKDSK88) [inline target/linux/layerscape/base-files/etc/board.d/02_network:25]: IMO, this part should be in [05_layerscape_reorder_eth](https://github.com/openwrt/openwrt/blob/master/target/linux/layerscape/base-files/lib/preinit/05_layerscape_reorder_eth).

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (CHKDSK88) [inline target/linux/layerscape/base-files/etc/board.d/02_network:26]: Is it necessary? Other devices with DSA doesn't up interface manually.

File: package/firmware/layerscape/ls-ddr-phy/Makefile
Comment (CHKDSK88) [reply package/firmware/layerscape/ls-ddr-phy/Makefile:1]: Please move it to first line:

https://spdx.github.io/spdx-spec/appendix-V-using-SPDX-short-identifiers-in-source-files/#format-for-spdx-license-identifier

**Review by CHKDSK88 — COMMENTED** [review]
@yangbolu1991 

Thank You for update. I mark some issues.

`bootm_size` works also with LS1012A-RDB and it's fine for me.

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (yangbolu1991) [reply target/linux/layerscape/base-files/etc/board.d/02_network]: Ok.

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (yangbolu1991) [reply target/linux/layerscape/base-files/etc/board.d/02_network]: It's necessary. This port is connecting to CPU port of the switch. It should be up always.

File: package/firmware/layerscape/ls-ddr-phy/Makefile
Comment (yangbolu1991) [reply package/firmware/layerscape/ls-ddr-phy/Makefile:2]: Ok.

**Comment by yangbolu1991** [comment]
Pushed v4 patch-set.
Changes for v4:
- Moved SPDX to first line.
- Moved interface re-naming to 05_layerscape_reorder_eth.

**Comment by yangbolu1991** [comment]
Any comments on the new version patches?
Thanks.

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (CHKDSK88) [inline target/linux/layerscape/base-files/etc/board.d/02_network:19]: This script isn't launched at every boot.

It should be moved to preinit scripts. Examples:

https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/lib/preinit/07_mt7621_bringup_dsa_master

https://github.com/openwrt/openwrt/blob/master/target/linux/mediatek/base-files/lib/preinit/05_set_preinit_iface

**Comment by adschm** [comment]
Picked "layerscape: make restool depend on TARGET_layerscape_armv8_64b".

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (yangbolu1991) [reply target/linux/layerscape/base-files/etc/board.d/02_network]: Done.

**Comment by yangbolu1991** [comment]
Pushed v5 patch-set.
Changes for v5:
- Moved "ip link set eno2 up" to preinit script.
- Rebased patches.

Thanks.

File: package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-uEnv.txt
Comment (CHKDSK88) [inline package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-uEnv.txt:3]: Please add bootm_size to consolidate envs because https://github.com/openwrt/openwrt/commit/65f8d7360c041ed90724bcc97673b30ce704cc33 was merged.

File: package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt
Comment (CHKDSK88) [inline package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt:4]: Please add bootm_size to consolidate envs because https://github.com/openwrt/openwrt/commit/65f8d7360c041ed90724bcc97673b30ce704cc33 was merged.

File: package/boot/uboot-layerscape/files/fsl_lx2160a-rdb-uEnv.txt
Comment (CHKDSK88) [inline package/boot/uboot-layerscape/files/fsl_lx2160a-rdb-uEnv.txt:4]: Please add bootm_size to consolidate envs because https://github.com/openwrt/openwrt/commit/65f8d7360c041ed90724bcc97673b30ce704cc33 was merged.

**Comment by CHKDSK88** [comment]
@yangbolu1991 

You need to rebase again. My PR was merged.

File: package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-uEnv.txt
Comment (yangbolu1991) [reply package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-uEnv.txt]: Done.

File: package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt
Comment (yangbolu1991) [reply package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt]: Done.

File: package/boot/uboot-layerscape/files/fsl_lx2160a-rdb-uEnv.txt
Comment (yangbolu1991) [reply package/boot/uboot-layerscape/files/fsl_lx2160a-rdb-uEnv.txt]: Done.

**Comment by yangbolu1991** [comment]
@CHKDSK88 
Good to see your pull request was merged. I rebased mine.

Changes for v6 patch-set,
- Rebased.
- Converted to use bootm_size env in u-boot for new added boards.

Thanks.

**Comment by yangbolu1991** [comment]
@adschm ,
Any comments on the new version?
Thanks.

File: package/firmware/layerscape/ls-ddr-phy/Makefile
Comment (yangbolu1991) [reply package/firmware/layerscape/ls-ddr-phy/Makefile:16]: Already addressed since v3.

**Comment by yangbolu1991** [comment]
@adschm ,
Will you merge the pull request?
Thanks.

**Comment by adschm** [comment]
> @adschm ,
> Will you merge the pull request?
> Thanks.

Sorry, probably not anytime soon. This is too big for my currently available time, and layerscape is actually not one of my primary platforms/realms, so reviewing it doesn't come easy. (That's why I picked some of the straightforward patches beforehand BTW).

But there are other committer as well, and for a big PR like this I fear you will just have to wait a little longer.

I might have another quick look this weak, though, maybe I can at least pick parts of it again; but no promises ...

**Comment by yangbolu1991** [comment]
@adschm ,
Rebased and build test. No code changes.
I understand. I think I need to change my upstreaming strategy next time:)
These patches were from NXP released project and were tested on the code base.
I had thought it's easier for community acceptance after they were tested by us and sent together.
But I should have thought this would bring large code changes for reviewing one time. Sorry.

For this pull reqeust, the new boards support patches introduced large code changes. The other patches are minor changes.
I'm responsible to address any problem of conflict, build, and bug on your code base. Please feel free to review and merge.
6e0ceca layerscape: add LS1028ARDB board support
fabd01c layerscape: add FRWY-LS1046A board support
c07448c layerscape: add LX2160ARDB (Rev2.0 silicon) board support
b55e321 layerscape: add ddr-phy package

Thanks.

File: package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt
Comment (adschm) [inline package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt:6]: Shouldn't that be 49m@0xf00000 due to dtb inclusion in image?
@yangbolu1991 @CHKDSK88 
Please just answer here and do not update/repush, as I've already picked stuff.

File: package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-uEnv.txt
Comment (adschm) [inline package/boot/uboot-layerscape/files/fsl_ls1028a-rdb-uEnv.txt:6]: same question as above: 49m@0xf00000 ?

File: target/linux/layerscape/base-files/etc/board.d/02_network
Comment (adschm) [inline target/linux/layerscape/base-files/etc/board.d/02_network:21]: This should be moved for proper sorting.

File: target/linux/layerscape/base-files/lib/preinit/05_layerscape_reorder_eth
Comment (adschm) [inline target/linux/layerscape/base-files/lib/preinit/05_layerscape_reorder_eth:21]: Sorting.

File: package/boot/uboot-layerscape/files/fsl_lx2160a-rdb-uEnv.txt
Comment (adschm) [inline package/boot/uboot-layerscape/files/fsl_lx2160a-rdb-uEnv.txt:8]: 49m@0xf00000 ?

**Comment by adschm** [comment]
This would be the next subset of patches I'd take:
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/layers

However, this still needs confirmation for the 48m/49m question, which is _not_ included yet.
I haven't build-tested yet, so maybe I'll update if there are problems.

File: package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt
Comment (CHKDSK88) [reply package/boot/uboot-layerscape/files/fsl_ls1046a-frwy-uEnv.txt:6]: It should be `49m@0xf00000`. Now sysupgrade for this devices are broken.

**Comment by yangbolu1991** [comment]
@adschm @CHKDSK88 ,
Thank you very much for pointing out the firmware partition issue. Yes. It should be 49m@0xf00000.
And the sorting problem in scripts needs fix too.

Should I wait for cherry-picked patches merged by mainline, and rebase this pull request with fix?
https://git.openwrt.org/?p=openwrt/staging/adrian.git;a=shortlog;h=refs/heads/layers

Or update the pull request now?
Thank you.

**Comment by adschm** [comment]
@yangbolu1991 If you think my patches are okay, I'd merge them first (just with the 49m changed), and then you could rebase your remaining stuff afterwards.

**Comment by yangbolu1991** [comment]
@adschm All your patches look good. I will wait for merging them, and rebase the remaining soon after that.
Thank you very much.

**Comment by adschm** [comment]
Patches merged.

**Comment by yangbolu1991** [comment]
Rebased the rest patches, with sorting problem and firmware address problem addressed.
Thanks a lot.

**Comment by CodeFetch** [comment]
@yangbolu1991 You have conflicting files. Why do you disable so many kernel symbols? Aren't they needed for something else?

**Pull request closed without merging**
